### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by MatthiasLohr** [Shoobx/xmldiff#4]
Hi,

i want to compare 2 2,2MB XML files, but xmldiif needs up to 8GB RAM for that - and gets then killed by my some security tools because of the RAM exhaust....

Is there any kind of mem leak? How to debug that?

Best regards
Matthias

**Comment by agroszer** [comment]
Hi,

Weird, my sources tell me that that file sizes should be no problem.

If you're on linux you can watch the processes memory usage with `top`. See if it skyrockets.

**Comment by MatthiasLohr** [comment]
```top``` is the toll with which i found the > 8GB RAM usage. But now idea how to have a deeper inspection.

**Comment by agroszer** [comment]
Are those XMLs private or can we have a look ourselves?

**Comment by agroszer** [comment]
google and SO spit out some results for "python memory profiler", like: https://stackoverflow.com/questions/110259/which-python-memory-profiler-is-recommended

**Comment by MatthiasLohr** [comment]
Attached the problematic XML files.

[xmldiff.tar.gz](https://github.com/Shoobx/xmldiff/files/2089952/xmldiff.tar.gz)

**Comment by fontealpina** [comment]
Same issue trying to compare two files with ~ 10k lines

**Comment by regebro** [comment]
Yeah, it looks like the lcs2() algorithm has cubic memory usage, I think that's the problem. It's certainly in an lcs2() call the issue is, and that goes for both the Python and the C implementation.

**Comment by fontealpina** [comment]
So if I well understand it is not strictly a bug but more a an algorithm limitation.  
Is there a way to overcome this?

**Comment by regebro** [comment]
I tried implementing an algorithm that could use integer arrays instead of Python lists, but that just means we run out of memory more slowly. So that would be an improvement, but the example XML files would still fail.

**Comment by wmahlous** [comment]
I would also be interested in a fix for this. 2 XML files which were 1.3MB each used up > 95 GB of memory (not my personal machine, I wish ...) and caused kernel OOM. The files are restricted so I cannot share them directly but in a few weeks when I get some time I will create similar ones with dummy data to reproduce.

**Comment by regebro** [comment]
That's OK, we know what the issue is, and have files to reproduce it. Because of a different issue, I'm going to implement an alternative slower algorithm during the next few days, I think it will use less memory, but I'll promise nothing. :-)

**Comment by regebro** [comment]
No, there IS some sort of memory leak in the C version. I've also done performance tests and the C version really isn't that much faster. I have a python-only branch which you might try: https://github.com/Shoobx/xmldiff/tree/lregebro-python-only

Diffing a file with nearly a 100.000 sax items isn't fast though, don't be surprised if it's hours.

**Comment by regebro** [comment]
In fact, I think xmldiff ends up in an infinite loop of some sort with those test files, although I'm not 100% sure yet. That + memory leak is probably the reason.

**Comment by MatthiasLohr** [comment]
Even if it's not a memory leak: An inifinte loop requesting memory for every loop will lead to the same effect. The memory is not given back to the system - which might be right, as long as it's running. But the "infinite running" might be wrong ;)

**Comment by algupta** [comment]
Is there any update on this please?

**Comment by regebro** [comment]
@wmahlous So on second thought, if you can create files I can test I would appreciate that.

**Comment by regebro** [comment]
@MatthiasLohr Does this diff look reasonable for you, for the example files you added?

[bigdiff.txt](https://github.com/Shoobx/xmldiff/files/2232112/bigdiff.txt)

**Comment by MatthiasLohr** [comment]
@regebro, I'm not sure how to read that file. That does't look like a typical diff file.

**Comment by regebro** [comment]
That's an output that isn't exactly what you would get from xmldiff, but very similar.

Each line is an action, like delete, or update, and you will get at least one xpath to what is to be deleted or updated or inserted, etc.

Here an actual diff output from xmldiff, for a smaller file:
[rmldiff.txt](https://github.com/Shoobx/xmldiff/files/2262506/rmldiff.txt)

**Comment by wmahlous** [comment]
@regebro Sorry for the late response, I was on holiday. Attached are 2 dummy XMLs that you can use. **terminal_output.txt** shows what happens when I limit the process to 10 GB. If I hadn't set that limit it would continue to climb until the host OOMs. Let me know if you can reproduce.

[dummy_xml.zip](https://github.com/Shoobx/xmldiff/files/2274766/dummy_xml.zip)

**Comment by wmahlous** [comment]
@regebro Sorry for the late response, was on holiday. Here's a dummy file that you can reproduce with: https://raw.githubusercontent.com/wmahlous/playground/master/dummy_xml_uPVzCTmZ.xml
In the same repo there's another file ending in "_2" which you can use as the second file. It has a few minor differences.

Here's a demo (setting 10GB limit to avoid OOM):

```
$ ulimit -v
unlimited

$ ulimit -Sv 10000000

$ ulimit -v
10000000

$ ls -hl dummy_xml_uPVzCTmZ.xml
-rw-r--r-- 1 abcd efgh 1.2M Aug  9 16:11 dummy_xml_uPVzCTmZ.xml

$ ls -hl dummy_xml_uPVzCTmZ_2.xml
-rw-r--r-- 1 abcd efgh 1.2M Aug  9 16:10 dummy_xml_uPVzCTmZ_2.xml

$ xmldiff dummy_xml_uPVzCTmZ.xml dummy_xml_uPVzCTmZ_2.xml
Memory fault(coredump)
```

Please let me know if you can also reproduce with the dummy files.

**Comment by wmahlous** [comment]
@regebro Sorry for the late response, was on holiday. Here's a dummy file that you can reproduce with: https://raw.githubusercontent.com/wmahlous/playground/master/dummy_xml_uPVzCTmZ.xml
In the same repo there's another file ending in "_2" which you can use as the second file. It has a few minor differences.

Here's a demo (setting 10GB limit to avoid OOM):

```
$ ulimit -v
unlimited

$ ulimit -Sv 10000000

$ ulimit -v
10000000

$ ls -hl dummy_xml_uPVzCTmZ.xml
-rw-r--r-- 1 abcd efgh 1.2M Aug  9 16:11 dummy_xml_uPVzCTmZ.xml

$ ls -hl dummy_xml_uPVzCTmZ_2.xml
-rw-r--r-- 1 abcd efgh 1.2M Aug  9 16:10 dummy_xml_uPVzCTmZ_2.xml

$ xmldiff dummy_xml_uPVzCTmZ.xml dummy_xml_uPVzCTmZ_2.xml
Memory fault(coredump)
```

Please let me know if you can also reproduce.

**Comment by regebro** [comment]
Yes, I can reproduce that, and I can confirm the the pure python version also never finishes, so it's the same issue.

I'm working on a modern python reimplementation of xmldiff. It gives the following editscrip / diff.
xmldiff compatible outputs is still on my TODO list:

```
[InsertNode(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct', tag='nxiDMhAvOfDcQx', position=5),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/nxiDMhAvOfDcQx', name='FNmhE', value='Ntpa'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/nxiDMhAvOfDcQx', name='afdg', value='TUTIBQCRZmAtuuAVeh'),
 InsertNode(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct', tag='XwjizmWsGKBDCO', position=6),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/XwjizmWsGKBDCO', name='Gpvwb', value='K'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/XwjizmWsGKBDCO', name='aqUn', value='JVFtSUrygRjgaypVL'),
 InsertNode(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct', tag='crfBuUTBDZJSwd', position=7),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/crfBuUTBDZJSwd', name='WiVgz', value='zJSGbapmkfqaFn'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/crfBuUTBDZJSwd', name='ZZMh', value='OUHIwtXuHRYg'),
 UpdateAttrib(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/mbUdQSpfOua', name='xVIClVgM', value='Mivj'),
 InsertNode(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS', tag='rBuHqfVEttS', position=19),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='TzCbHSlkK', value='a'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='UHLG', value='siqqHRxcLEvAsmWORGU'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='UoOJGnoL', value='SWINcDV'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='bojksUmfuhRE', value=''),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='hrELoAHa', value='OBdV'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='jpqdO', value='B'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='srEDSJcAvqO', value='bB'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='tiUIINLsQRQ', value=''),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='vyRvxgJaDRr', value=''),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', name='ykSlPWw', value='SGyqYMyoS'),
 InsertNode(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS', tag='PrRmPPPWXaM', position=20),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='CvNCYJeSujU', value=''),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='KJkoyWlQ', value='BVYd'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='OppWMyc', value='SpOgpRIaK'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='RvYiP', value='u'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='WSzRTKtgnJDx', value=''),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='YDRjxbdTYpu', value=''),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='pjUZqdwUD', value='S'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='vNce', value='OvfVYOCcozUmVDfuSIPu'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='vQGFUbDdmNB', value='BD'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', name='wQFMKMmA', value='EpOlGlm'),
 InsertNode(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/HJarmtER', tag='KVPgZsTID', position=25),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/HJarmtER/KVPgZsTID', name='VTLB', value='NPUZUpNeIJGmsnBFyaHZHckJqjFUKDOYfMgAQpopPGa'),
 InsertAttrib(target='/ffSYlxhrW/tLqXefNXjs/MiizZf/HJarmtER/KVPgZsTID', name='hlrUe', value='xw'),
 DeleteNode(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/HJarmtER/KVPgZsTIDWAEL'),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/CKjBlwzcDdGjDY', text='\n        '),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/nxiDMhAvOfDcQx', text='\n        '),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/XwjizmWsGKBDCO', text='\n        '),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/VJAcct/crfBuUTBDZJSwd', text='\n    '),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/rBuHqfVEttS', text='\n        '),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/GjwynS/PrRmPPPWXaM', text='\n        '),
 UpdateTextAfter(node='/ffSYlxhrW/tLqXefNXjs/MiizZf/HJarmtER/KVPgZsTID', text='\n        ')]
```

**Comment by regebro** [comment]
I've just released a beta version of xmldiff 2.0: https://pypi.org/project/xmldiff/2.0b1/
It can diff these files. Feedback is appreciated.
